default.nix

```diff
@@ -0,0 +1,277 @@
+
+#  nixos-mailserver: a simple mail server
+#  Copyright (C) 2016-2017  Robin Raymond
+#
+#  This program is free software: you can redistribute it and/or modify
+#  it under the terms of the GNU General Public License as published by
+#  the Free Software Foundation, either version 3 of the License, or
+#  (at your option) any later version.
+#
+#  This program is distributed in the hope that it will be useful,
+#  but WITHOUT ANY WARRANTY; without even the implied warranty of
+#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#  GNU General Public License for more details.
+#
+#  You should have received a copy of the GNU General Public License
+#  along with this program. If not, see <http://www.gnu.org/licenses/>
+
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.mailserver;
+in
+{
+  options.mailserver = {
+    enable = mkEnableOption "nixos-mailserver";
+
+    domain = mkOption {
+      type = types.str;
+      example = "example.com";
+      description = "The domain that this mail server serves. So far only one domain is supported";
+    };
+
+    host_prefix = mkOption {
+      type = types.str;
+      default = "mail";
+      description = ''
+        The prefix of the FQDN of the server. In this example the FQDN of the server
+        is given by 'mail.example.com'
+      '';
+    };
+
+    login_accounts = mkOption {
+      type = types.loaOf (types.submodule ({ name, ... }: {
+        options = {
+          name = mkOption {
+            type = types.str;
+            example = "user1";
+            description = "Username";
+          };
+
+          hashedPassword = mkOption {
+            type = types.str;
+            example = "$6$evQJs5CFQyPAW09S$Cn99Y8.QjZ2IBnSu4qf1vBxDRWkaIZWOtmu1Ddsm3.H3CFpeVc0JU4llIq8HQXgeatvYhh5O33eWG3TSpjzu6/";
+            description = ''
+              Hashed password. Use `mkpasswd` as follows
+
+              ```
+              mkpasswd -m sha-512 "super secret password"
+              ```
+            '';
+          };
+        };
+
+        config.name = mkDefault name;
+      }));
+      example = {
+        user1 = {
+          hashedPassword = "$6$evQJs5CFQyPAW09S$Cn99Y8.QjZ2IBnSu4qf1vBxDRWkaIZWOtmu1Ddsm3.H3CFpeVc0JU4llIq8HQXgeatvYhh5O33eWG3TSpjzu6/";
+        };
+        user2 = {
+          hashedPassword = "$6$oE0ZNv2n7Vk9gOf$9xcZWCCLGdMflIfuA0vR1Q1Xblw6RZqPrP94mEit2/81/7AKj2bqUai5yPyWE.QYPyv6wLMHZvjw3Rlg7yTCD/";
+        };
+      };
+      description = ''
+        The login account of the domain. Every account is mapped to a unix user,
+        e.g. `user1@example.com`. To generate the passwords use `mkpasswd` as
+        follows
+
+        ```
+        mkpasswd -m sha-512 "super secret password"
+        ```
+      '';
+    };
+
+    valiases = mkOption {
+      type = types.attrsOf (types.enum (builtins.attrNames cfg.login_accounts));
+      example = {
+        info = "user1";
+        postmaster = "user1";
+        abuse = "user1";
+      };
```

Comment:
I feel like this form of declaration makes it hard to extend it to multiple domains. Is `info@example.com` a legal identifier in a `Nix` set?

Reply:
I didn't know you intended to support multiple domains (since the `domain` option is a single item), but yes, any string can be used as an identifier if you quote it: `"info@example.com"`.

Reply:
Ok, then this should not be a problem.